[BugFix] update error when table with generate column (#25600)

srlch opened the pull request:
Fix the following Problems:
1. The variable selectlist in the update analyzer miss for generated column causing NPE problem
2. The update stmt with partial update mode will write the partial segment file
    and does not need all slotDescriptor except for generated column.
    Because generated can not be set in the assignment list but it must 
    be computed and written in the OlapTableSink phase.
4. partial update for update stmt could not miss all ref columns for all generated columns.

Fixes #25600

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 22 (04.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|0|20|00.00%|[154, 181, 182, 183, 184, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 200, 201, 204, 205]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|1|2|50.00%|[88]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 22 (04.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|0|20|00.00%|[154, 181, 182, 183, 184, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 200, 201, 204, 205]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|1|2|50.00%|[88]|

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/sql/UpdatePlanner.java:
```diff
@@ -84,8 +84,10 @@ public ExecPlan plan(UpdateStmt updateStmt, ConnectContext session) {
             for (Column column : table.getFullSchema()) {
-                if (updateStmt.usePartialUpdate() && !updateStmt.isAssignmentColumn(column.getName()) && !column.isKey()) {
-                    // When using partial update, skip columns which aren't key column and not be assign
+                if (updateStmt.usePartialUpdate() && !column.isMaterializedColumn() &&
```
What's this meaning?
The generated column can be not written in the update DML?

srlch replied:
In partial update mode, generated column must be computed in OlapTableSink phase and can not be done in apply. The code here means that we SHOULD computed and write the generated column in the OlapTableSink phase. And it does not means that we could rewritten the  generated column using some specified value by user. e.g. update t set generated_column = value where pk = xxx will always throw an exception.

srlch committed:
[BugFix] update error when table with generate column (#25600)

Fix the following Porblems:
1. The variable selectlist in update anayzler miss for generated column cause NPE problem
2. The update stmt with partial update mode will write the partial segment file and do not need all slotDescriptor except for generated column. Because generated can not be set in assignment list but it must be computed and write in OlapTableSink phase.
3. partial update for update stmt could not miss all ref column for all generated column.

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 22 (04.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|0|20|00.00%|[154, 181, 182, 183, 184, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 200, 201, 204, 205]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|1|2|50.00%|[88]|

wanpengfei-git commented:
@Mergifyio backport branch-3.1
